h2o-core/src/main/java/water/rapids/SortCombine.java

```diff
@@ -0,0 +1,407 @@
+    private final Frame _frame;
+    private final Vec _vec;
+    private final int _chunkNode[]; // Chunk homenode index
+    final int _msb;
+
+    FFSB(Frame frame, int msb) {
+      assert -1 <= msb && msb <= 255; // left ranges from 0 to 255, right from -1 to 255
+      _frame = frame;
+      _msb = msb;
+      
+      // Create fast lookups to go from chunk index to node index of that chunk
+      Vec vec = _vec = frame.anyVec();
+      _chunkNode = vec == null ? null : MemoryManager.malloc4(vec.nChunks());
+      if (vec == null) return; // Zero-columns for Sort
+      for (int i = 0; i < _chunkNode.length; i++)
+        _chunkNode[i] = vec.chunkKey(i).home_node().index();
+    }
+  }
+  
+  SortCombine(FFSB leftSB, SingleThreadRadixOrder.OXHeader leftSortedOXHeader) {
+    _leftSB = leftSB;
+    int columnsInResult = _leftSB._frame.numCols();
+    _stringCols = MemoryManager.mallocZ(columnsInResult);
+    _intCols = MemoryManager.mallocZ(columnsInResult);
+    if (_leftSB._frame!=null) {
+      for (int col=0; col < _leftSB._frame.numCols(); col++) {
+        if (_leftSB._frame.vec(col).isInt())
+          _intCols[col] = true;
+        if (_leftSB._frame.vec(col).isString())
+          _stringCols[col] = true;
+      }
+    }
+    _chunkNum = _leftSB._frame.anyVec().nChunks();
+    _leftSortedOXHeader = leftSortedOXHeader;
+  }
+
+  @Override
+  public void compute2() {
+    _timings = MemoryManager.malloc8d(20);
+    long t0 = System.nanoTime();
+    _leftKO = new KeyOrder(_leftSortedOXHeader);
+    _leftKO.initKeyOrder(_leftSB._msb,/*left=*/true);
+    final long leftN = _leftSortedOXHeader._numRows; // store number of rows in left frame for the MSB
+    assert leftN >= 1;
+    _timings[0] += (System.nanoTime() - t0) / 1e9;
+    
+    _leftFrom = -1;
+    long leftTo = leftN; // number of rows in left frame
+
+    long retSize = leftTo - _leftFrom - 1;
+    assert retSize >= 0;
+    if (retSize == 0) {
+      tryComplete();
+      return;
+    }
+
+    _retBatchSize = 1048576;   // must set to be the same from RadixOrder.java
+    _numRowsInResult = retSize;
+    
+    setPerNodeNumsToFetch();  // find out the number of rows to fetch from H2O nodes, number of rows to fetch per chunk
+    
+    if (_numRowsInResult > 0) createChunksInDKV();
+    tryComplete();
+  }
+
+  /**
+   * This method will find the number of rows per chunk to fetch per batch for a MSB
+   */
+  private void setPerNodeNumsToFetch() {
+    Vec anyVec = _leftSB._frame.anyVec();
+    int nbatch = _leftKO._order.length;
+    _numRowsPerCidx = new int[nbatch][anyVec.nChunks()];
+    for (int batchNum =0; batchNum < nbatch; batchNum++) {
+      int batchSize = _leftKO._order[batchNum].length;
+      for (int index=0; index < batchSize; index++) {
+        long globalRowNumber = _leftKO._order[batchNum][index];
+        int chkIdx = _leftSB._vec.elem2ChunkIdx(globalRowNumber);
+        _leftKO._perNodeNumRowsToFetch[_leftSB._chunkNode[chkIdx]]++;
```

Comment:
we go over the whole dataframe once here in order to discuss for each batch, how many rows per chunks need to be fetched to form the final sorted chunks for that MSB